File: NavieraCruzApp/Modules/Fleet/FleetView.swift
import SwiftUI
import AVKit
import AVFoundation

struct FleetView: View {
    @StateObject private var viewModel = FleetViewModel()
    
    var body: some View {
        NavigationView {
            ZStack {
                ColorTheme.secondaryBackground.ignoresSafeArea()
                
                if viewModel.isLoading && viewModel.ships.isEmpty {
                    ProgressView()
                } else if viewModel.ships.isEmpty {
                    Text("No hay barcos disponibles.")
                } else {
                    ScrollView {
                        LazyVStack(spacing: 16) {
                            ForEach(viewModel.ships) { ship in
                                ShipCardView(ship: ship)
                            }
                        }
                        .padding()
                    }
                    .refreshable {
                        viewModel.loadShips()
                    }
                }
            }
            .navigationTitle("Dashboard de Flota")
            .onAppear {
                viewModel.loadShips()
            }
        }
    }
}

struct ShipCardView: View {
    let ship: Ship
    
    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack {
                Text(ship.name)
                    .font(Typography.headline())
                    .foregroundColor(ColorTheme.textPrimary)
                
                Spacer()
                
                Text(ship.status.rawValue)
                    .font(.caption)
                    .fontWeight(.bold)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(statusColor(ship.status).opacity(0.2))
                    .foregroundColor(statusColor(ship.status))
                    .cornerRadius(8)
            }
            
            Divider()
            
            HStack {
                VStack(alignment: .leading) {
                    Text("Carga Total")
                        .font(.caption)
                        .foregroundColor(ColorTheme.textSecondary)
                    Text("\(Int(ship.totalCargo)) tons")
                        .font(Typography.body())
                        .fontWeight(.semibold)
                }
                Spacer()
                VStack(alignment: .trailing) {
                    Text("Ubicación")
                        .font(.caption)
                        .foregroundColor(ColorTheme.textSecondary)
                    Text("Lat: \(String(format: "%.2f", ship.latitude)), Lon: \(String(format: "%.2f", ship.longitude))")
                        .font(.caption)
                }
            }
            
            // Functional Video Player for SBS Camera (AVKit HLS Player)
            if let cameraUrlStr = ship.cameraUrl, let url = URL(string: cameraUrlStr) {
                VideoPlayer(player: AVPlayer(url: url))
                    .frame(height: 150)
                    .cornerRadius(8)
            } else {
                ZStack {
                    Rectangle()
                        .fill(Color.gray.opacity(0.1))
                        .frame(height: 150)
                        .cornerRadius(8)
                    
                    VStack {
                        Image(systemName: "video.slash.fill")
                            .foregroundColor(.gray)
                        Text("Cámara SBS no disponible")
                            .font(.caption)
                            .foregroundColor(.gray)
                    }
                }
            }
            .padding(.top, 8)
        }
        .padding()
        .background(ColorTheme.background)
        .cornerRadius(12)
        .shadow(color: Color.black.opacity(0.05), radius: 5, x: 0, y: 2)
    }
    
    func statusColor(_ status: ShipStatus) -> Color {
        switch status {
        case .active: return .green
        case .maintenance: return .orange
        case .docked: return .blue
        }
    }
}

File: naviera_cruz_app/lib/modules/fleet/fleet_view.dart
import 'package:flutter/material.dart';
import '../../models/models.dart';
import '../../services/services.dart';
import '../../app/theme.dart';
import '../../app/bar_widget.dart';
import '../crew/crew_list_view.dart';
import 'sbs_camera_player.dart';

class FleetView extends StatefulWidget {
  const FleetView({super.key});

  @override
  State<FleetView> createState() => _FleetViewState();
}

class _FleetViewState extends State<FleetView> {
  final List<Ship> _ships = [];
  bool _isLoading = false;
  String? _errorMessage;
  int _selectedShipIndex = 0;

  @override
  void initState() {
    super.initState();
    _loadShips();
  }

  Future<void> _loadShips() async {
    setState(() {
      _isLoading = true;
      _errorMessage = null;
    });

    try {
      final fleetService = FleetService();
      final ships = await fleetService.fetchShips();
      setState(() {
        _ships.clear();
        _ships.addAll(ships);
        // Ensure index is within range
        if (_selectedShipIndex >= _ships.length) {
          _selectedShipIndex = 0;
        }
      });
    } catch (e) {
      setState(() {
        _errorMessage = e.toString();
      });
    } finally {
      setState(() {
        _isLoading = false;
      });
    }
  }

  Color _statusColor(ShipStatus status) {
    switch (status) {
      case ShipStatus.active: return Colors.green;
      case ShipStatus.maintenance: return Colors.orange;
      case ShipStatus.docked: return ColorTheme.primary;
    }
  }

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);

    return Scaffold(
      appBar: const NavieraAppBar(),
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          // Title
          Padding(
            padding: const EdgeInsets.fromLTRB(20.0, 16.0, 20.0, 8.0),
            child: Text(
              "Flota y operaciones",
              style: TextStyle(
                color: ColorTheme.primary,
                fontSize: 24,
                fontWeight: FontWeight.bold,
                letterSpacing: -0.5,
              ),
            ),
          ),
          
          // Ship Selection Tabs (Chips)
          if (!_isLoading && _ships.isNotEmpty)
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 20.0, vertical: 8.0),
              child: SingleChildScrollView(
                scrollDirection: Axis.horizontal,
                child: Row(
                  children: _ships.asMap().entries.map((entry) {
                    final index = entry.key;
                    final ship = entry.value;
                    final isSelected = _selectedShipIndex == index;
                    
                    Color chipBgColor = Colors.grey.shade400;
                    if (isSelected) {
                      chipBgColor = ColorTheme.primary;
                    }

                    return Padding(
                      padding: const EdgeInsets.only(right: 10.0),
                      child: GestureDetector(
                        onTap: () {
                          setState(() {
                            _selectedShipIndex = index;
                          });
                        },
                        child: Container(
                          padding: const EdgeInsets.symmetric(horizontal: 18, vertical: 10),
                          decoration: BoxDecoration(
                            color: chipBgColor,
                            borderRadius: BorderRadius.circular(20),
                          ),
                          child: Text(
                            ship.name,
                            style: const TextStyle(
                              color: Colors.white,
                              fontWeight: FontWeight.bold,
                              fontSize: 14,
                            ),
                          ),
                        ),
                      ),
                    );
                  }).toList(),
                ),
              ),
            ),
          
          Expanded(
            child: _isLoading && _ships.isEmpty
                ? const Center(child: CircularProgressIndicator())
                : RefreshIndicator(
                    onRefresh: _loadShips,
                    child: _ships.isEmpty
                        ? Center(
                            child: SingleChildScrollView(
                              physics: const AlwaysScrollableScrollPhysics(),
                              padding: const EdgeInsets.all(40.0),
                              child: Text(
                                _errorMessage ?? "No hay barcos disponibles.",
                                style: const TextStyle(color: Colors.grey),
                                textAlign: TextAlign.center,
                              ),
                            ),
                          )
                        : ListView(
                            physics: const AlwaysScrollableScrollPhysics(),
                            padding: const EdgeInsets.symmetric(horizontal: 20.0, vertical: 8.0),
                            children: [
                              _buildShipCard(context, _ships[_selectedShipIndex], theme),
                            ],
                          ),
                  ),
          ),
        ],
      ),
    );
  }

  Widget _buildShipCard(BuildContext context, Ship ship, ThemeData theme) {
    final statusCol = _statusColor(ship.status);
    final isDark = theme.brightness == Brightness.dark;
    final textColor = isDark ? Colors.white : Colors.black87;
    final secondaryTextColor = isDark ? Colors.white70 : Colors.black54;
    final dividerColor = isDark ? Colors.white.withOpacity(0.1) : const Color(0xFFF1F5F9);

    return Card(
      margin: const EdgeInsets.only(bottom: 16.0),
      child: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Text(
                  ship.name,
                  style: TypographyTheme.headline(context).copyWith(
                    color: textColor,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                Container(
                  padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 5),
                  decoration: BoxDecoration(
                    color: statusCol.withOpacity(0.1),
                    borderRadius: BorderRadius.circular(12),
                    border: Border.all(color: statusCol.withOpacity(0.3), width: 0.8),
                  ),
                  child: Text(
                    ship.status.rawValue,
                    style: TextStyle(
                      color: statusCol,
                      fontSize: 11,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                ),
              ],
            ),
            Divider(height: 24, color: dividerColor),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text(
                      "Carga Total",
                      style: TypographyTheme.caption(context).copyWith(color: secondaryTextColor),
                    ),
                    const SizedBox(height: 4),
                    Text(
                      "${ship.totalCargo.toInt()} tons",
                      style: theme.textTheme.bodyLarge?.copyWith(
                        fontWeight: FontWeight.bold,
                        color: textColor,
                      ),
                    ),
                  ],
                ),
                Column(
                  crossAxisAlignment: CrossAxisAlignment.end,
                  children: [
                    Text(
                      "Ubicación",
                      style: TypographyTheme.caption(context).copyWith(color: secondaryTextColor),
                    ),
                    const SizedBox(height: 4),
                    Text(
                      "Lat: ${ship.latitude.toStringAsFixed(2)}, Lon: ${ship.longitude.toStringAsFixed(2)}",
                      style: TextStyle(
                        fontSize: 12,
                        color: secondaryTextColor,
                        fontWeight: FontWeight.w500,
                      ),
                    ),
                  ],
                ),
              ],
            ),
            const SizedBox(height: 16),
            
            // Video Player or Fallback for SBS camera
            if (ship.cameraUrl != null && ship.cameraUrl!.isNotEmpty)
              SBSCameraPlayer(
                key: ValueKey(ship.cameraUrl),
                url: ship.cameraUrl!,
              )
            else
              Container(
                height: 150,
                decoration: BoxDecoration(
                  color: Colors.grey.withOpacity(0.05),
                  borderRadius: BorderRadius.circular(12),
                  border: Border.all(color: const Color(0xFFE2E8F0)),
                ),
                child: Center(
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      const Icon(Icons.videocam_off_outlined, color: Colors.grey, size: 30),
                      const SizedBox(height: 6),
                      Text(
                        "Cámara SBS no disponible",
                        style: TypographyTheme.caption(context),
                      ),
                    ],
                  ),
                ),
              ),
            
            const SizedBox(height: 16),
            SizedBox(
              width: double.infinity,
              height: 44,
              child: OutlinedButton.icon(
                style: OutlinedButton.styleFrom(
                  side: const BorderSide(color: ColorTheme.primary),
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(22),
                  ),
                ),
                onPressed: () {
                  Navigator.push(
                    context,
                    MaterialPageRoute(
                      builder: (_) => CrewListView(shipId: ship.id),
                    ),
                  );
                },
                icon: const Icon(Icons.people_outline_outlined, color: ColorTheme.primary),
                label: const Text(
                  "Ver Tripulación",
                  style: TextStyle(
                    color: ColorTheme.primary,
                    fontWeight: FontWeight.bold,
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}
